File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    
    void solve(vector<vector<string>> &ans, vector<string> &pal, string s, int idx){
        if(s.size() == idx){
            ans.push_back(pal);
            return;
        }
        
        
        for(int i = idx; i<s.size(); i++){
            if(check(s,idx,i)){
                pal.push_back(s.substr(idx, i-idx+1));
                solve(ans, pal, s, i+1);
                pal.pop_back();
            }
        }
    }
    
    bool check(string s, int start, int last){
        while(start <= last ){
            if(s[start++] != s[last--])
                return false;
        }
        return true;
    }
    
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> pal;
        solve(ans, pal, s, 0);
        return ans;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    
    public void palindrome(int idx, String s, List<List<String>> ans, List<String> res){
        if(idx == s.length()){
            ans.add(new ArrayList<>(res));
        }
        
        for(int i = idx; i < s.length(); i++){
            if(isPalindrome(s,idx,i)){
                res.add(s.substring(idx, i+1));
                palindrome(i+1,s,ans,res);
                res.remove(res.size()-1);
            }
        }
    }
    
    public boolean isPalindrome(String s, int i, int j){
        while(i <= j){
            if(s.charAt(i++) != s.charAt(j--)){
                return false;
            }
        }
        return true;
    }
    
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();
        List<String> res = new ArrayList<>();
        
        palindrome(0,s,ans,res);
        
        return ans;
    }
}